src/Native/System.Native/pal_console.cpp

```diff
@@ -0,0 +1,231 @@
+            // Regardless, no more data is available, so we close the read
+            // end of the pipe and exit.
+            close(pipeFd);
+            return nullptr;
+        }
+
+        // Invoke the callback.  We take the lock while calling it so as to prevent
+        // it from being removed while we're still using it.
+        CtrlCallback callback = g_ctrlCallback;
+        int rv = callback != nullptr ?  callback(signalCode == SIGQUIT ? Break : Interrupt) : 0;
+        if (rv == 0) // callback removed or was invoked and didn't handle the signal
+        {
+             // The proper behavior for unhandled SIGINT/SIGQUIT is to set the signal 
+             // handler to the default and then send the signal to self and let the 
+             // default handler do its work.
+             struct sigaction resetAction = {
+                 .sa_handler = SIG_DFL,
+                 .sa_flags = 0
+             };
+             sigemptyset(&resetAction.sa_mask);
+
+             sigaction(SIGINT, &resetAction, nullptr);
+             sigaction(SIGQUIT, &resetAction, nullptr);
+
+             kill(getpid(), signalCode);
+         }
+    }
+}
+
+extern "C" int32_t RegisterForCtrl(CtrlCallback callback)
+{
+    assert(callback != nullptr);
+    assert(g_ctrlCallback == nullptr);
+
+    // Create a pipe we'll use to communicate with our worker
+    // thread.  We can't do anything interesting in the signal handler,
+    // so we instead send a message to another thread that'll do
+    // the handling work.
+    if (pipe(g_signalPipe) != 0)
+    {
+        return 0;
+    }
+    assert(g_signalPipe[0] >= 0);
+    assert(g_signalPipe[1] >= 0); 
+
+    // Create a small object to pass the read end of the pipe to the worker.
+    int* readFdPtr = reinterpret_cast<int*>(malloc(sizeof(int)));
+    if (readFdPtr == nullptr)
+    {
+        close(g_signalPipe[0]);
+        close(g_signalPipe[1]);
+        errno = ENOMEM;
+        return 0;
+    }
+    *readFdPtr = g_signalPipe[0];
+
+    // The pipe is created.  Create the worker thread.
+    pthread_t handlerThread;
+    if (pthread_create(&handlerThread, nullptr, CtrlHandleLoop, readFdPtr) != 0)
+    {
+        int err = errno;
+        free(readFdPtr);
+        close(g_signalPipe[0]);
+        close(g_signalPipe[1]);
+        errno = err;
+        return 0;
+    }
+
+    // Now that all of the state is configured, store our callback.
+    // This must be set prior to calling sigaction.
+    g_ctrlCallback = callback;
+
+    // Finally, register the signal handlers for SIGINT and SIGQUIT.
+    struct sigaction newAction = {
+        .sa_flags = SA_RESTART,
+        .sa_handler = &sigintquit_handler,
+    };
+    sigemptyset(&newAction.sa_mask);
+
+    int rv = sigaction(SIGINT, &newAction, &g_previousSigIntAction);
+    assert(rv == 0);
+    rv = sigaction(SIGQUIT, &newAction, &g_previousSigQuitAction);
+    assert(rv == 0);
+
+    // Success
+    return 1;
+}
+
+extern "C" void UnregisterForCtrl()
```

Comment:
I am not sure whether it is a good idea to ever unregister. It just makes interesting race condition (e.g. unregister, immediately register again while the signal handling thread for the old registration have not finished running yet - there is a chance that you will end up with two signal handling threads active...).

The current CoreCLR implementation does not unregisters the signal handler either.